### PR TITLE
tree-building : implement exercise

### LINKED ISSUE #741: tree-building: implement exercise
There are exercises that are not yet implemented in Python track (the full list of them - https://github.com/exercism/python/issues/417#issuecomment-334841778).
You can help us change this situation by implementing this one. Ref #417

The guide ["Implement an exercise from specification"](https://github.com/exercism/docs/blob/master/you-can-help/implement-an-exercise-from-specification.md) will help you succeed.

If you still have questions, I encourage you to ask them before starting working on the issue.

I also suggest claiming the issue by creating an empty PR as stated [here](https://github.com/exercism/docs/blob/master/you-can-help/implement-an-exercise-from-specification.md#avoiding-duplicate-work)

Please note that there are some special ['closing issue' keywords](https://help.github.com/articles/closing-issues-using-keywords/) that you can put in description of your PR and Github will take care of closing the corresponding issue automatically after merge.
This simple thing helps us a lot, and makes our life a bit easier

### PR DESCRIPTION
**Pull request by clapmyhands**
resolves #741

### CODE REVIEW

File: exercises/tree-building/example.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+class Node():
+    def __init__(self, node_id):
+        self.node_id = node_id
+        self.children = []
+
+
+def validateRecord(record):
+    if record.equal_id() and record.record_id != 0:
+        raise ValueError
```
Comment (m-a-ge) [inline exercises/tree-building/example.py]: Should probably provide the cause of the exception

File: exercises/tree-building/example.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    elif (not record.equal_id()) and record.parent_id >= record.record_id:
+        raise ValueError
```
Comment (m-a-ge) [inline exercises/tree-building/example.py]: the same

File: exercises/tree-building/example.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+class Node():
+    def __init__(self, node_id):
+        self.node_id = node_id
+        self.children = []
+
+
+def validateRecord(record):
+    if record.equal_id() and record.record_id != 0:
+        raise ValueError
+    elif (not record.equal_id()) and record.parent_id >= record.record_id:
```
Comment (m-a-ge) [inline exercises/tree-building/example.py]: redundant parentheses

**Comment by clapmyhands** [comment]
@m-a-ge I've change it according to the review. Please check again and let me know if there's anything else to change 😄 .

File: exercises/tree-building/tree_building_test.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            Record(0, 0),
+            Record(2, 0),
+            Record(1, 2)
+        ]
+        with self.assertRaises(ValueError):
+            BuildTree(records)
+
+    def assert_node_is_branch(self, node, node_id, children_count):
+        self.assertEqual(node.node_id, node_id)
+        self.assertFalse(len(node.children) == 0)
```
Comment (cmccandless) [inline exercises/tree-building/tree_building_test.py]: `assertNotEqual(len(node.children), 0)` would be clearer.

File: exercises/tree-building/tree_building_test.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        self.assertEqual(len(node.children), children_count)
+
+    def assert_node_is_leaf(self, node, node_id):
+        self.assertEqual(node.node_id, node_id)
+        self.assertTrue(len(node.children) == 0)
```
Comment (cmccandless) [inline exercises/tree-building/tree_building_test.py]: `assertEqual(len(node.children), 0)` would be clearer.

**Comment by clapmyhands** [comment]
@cmccandless I've changed according to your review. Also I've been thinking, do I need to put some reason on the test where a `ValueError` need to be raised?

**Comment by cmccandless** [comment]
@clapmyhands If I understand you correctly, do you mean something like this?
``` Python
with self.assertRaisesRegexp(ValueError, 'No root node'):
    BuildTree(records)
```

If that's what you mean, that would be a good idea :+1:

**Comment by clapmyhands** [comment]
@cmccandless I don't think it's possible. At least in my current code it can't since the checking of 'no root node' is the same as checking of 'non continuous'

reason being:
```python
raise ValueError
```
 can cover multiple failing cases.

I was thinking more in line of comments like so:
```python
# no root node
with self.assertRaises(ValueError):
            BuildTree(records)
```

the name of the test might already be self-explanatory but for some tests such as
```python
def test_cycle_directly(self):
```
it might be good to have further explanation on why that gives an error.

p.s: what do you think about changing
```python
self.assertIs(root, None)
```
to
```python
self.assertIsNone(root)
```

**Comment by clapmyhands** [comment]
@cmccandless I've changed it to use `self.assertIsNone` and added comment where I think is needed. Please review once more and let me know if there's anything else I need to change 😄

**Comment by cmccandless** [comment]
@clapmyhands I've left a couple comments. Just some consistency stuff.

**Comment by clapmyhands** [comment]
@cmccandless I've added comment for consistency.

**Comment by m-a-ge** [comment]
Thanks for working on this, @clapmyhands!

**Comment by clapmyhands** [comment]
@m-a-ge @cmccandless Thanks for being patient with me through it all. New contributor and hopefully a lasting one 😄

**Pull request merged**